mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs

```diff
@@ -1710,6 +1865,419 @@ public override string ToString ()
+		{
+			address.sa_family = sa_family;
+			address.SetDynamicLength (len);
+		}
+	};
+
+	// This is a class which can store arbitrary sockaddrs, even if they are not known the the Mono.Unix wrapper or the family does not have a corresponding value in the UnixAddressFamily enumeration.
+	[CLSCompliant (false)]
+	public sealed class SockaddrStorage : Sockaddr, IEquatable<SockaddrStorage> {
+		// Note: The sa_family field is ignored when passing a SockaddrStorage to a syscall (but it will be set when a SockaddrStorage is returned from a syscall). Instead of the sa_family field the value embedded in data is used.
+		public byte[] data;
+		public long data_len;
+
+		internal override byte[] DynamicData ()
+		{
+			return data;
+		}
+		internal override long GetDynamicLength ()
+		{
+			return data_len;
+		}
+		internal override void SetDynamicLength (long value)
+		{
+			data_len = value;
+		}
+
+		[DllImport (Syscall.MPH, SetLastError=true,
+				EntryPoint="Mono_Posix_SockaddrStorage_get_size")]
+		static extern int get_size ();
+		static readonly int default_size = get_size ();
+
+		public SockaddrStorage () : base (SockaddrType.SockaddrStorage | SockaddrType.MustBeWrapped, UnixAddressFamily.AF_UNSPEC)
+		{
+			data = new byte[default_size];
+			data_len = 0;
+		}
+		public SockaddrStorage (int size) : base (SockaddrType.SockaddrStorage | SockaddrType.MustBeWrapped, UnixAddressFamily.AF_UNSPEC)
+		{
+			data = new byte[size];
+			data_len = 0;
+		}
+
+		public unsafe void SetTo (Sockaddr address)
+		{
+			if (address == null)
+				throw new ArgumentNullException ("address");
+
+			var size = address.GetNativeSize ();
+			if (size > data.Length)
+				data = new byte[size];
+			fixed (byte* ptr = data)
+				if (!NativeConvert.TryCopy (address, (IntPtr) ptr))
+					throw new ArgumentException ("address", "Failed to convert to native struct");
+			data_len = size;
+			sa_family = address.sa_family;
+		}
+
+		public unsafe void CopyTo (Sockaddr address)
+		{
+			if (address == null)
+				throw new ArgumentNullException ("address");
+			if (data_len < 0 || data_len > data.Length)
+				throw new ArgumentException ("this", "data_len < 0 || data_len > data.Length");
+
+			fixed (byte* ptr = data)
+				if (!NativeConvert.TryCopy ((IntPtr) ptr, data_len, address))
+					throw new ArgumentException ("this", "Failed to convert from native struct");
+		}
+
+		public override string ToString ()
+		{
+			var sb = new StringBuilder ();
+			sb.AppendFormat ("{{sa_family={0}, data_len={1}, data=(", sa_family, data_len);
+			for (int i = 0; i < data_len; i++) {
+				if (i != 0)
+					sb.Append (" ");
+				sb.Append (data[i].ToString ("x2"));
+			}
+			sb.Append (")");
+			return sb.ToString ();
+		}
+
+		public override int GetHashCode ()
+		{
+			unchecked {
+				int hash = 0x1234;
+				for (int i = 0; i < data_len; i++)
+					hash += i ^ data[i];
+				return hash;
+			}
+		}
+		public override bool Equals (object obj)
+		{
+			if (!(obj is SockaddrStorage))
```

Comment:
Not necessary to change here, but for future information the `is` + cast pattern on value types can instead use `as` with a nullable type:

    var other = obj as SockaddrStorage?;
    if (other.HasValue)
        return Equals (other.Value);
    return false;